File: PythonCode/Chapter8_NonHermitian/Exercise_8_2_4.py
"""
 This script determines the admissible energies for a bound quantum
 particle confined in a Gaussian confining potential. It does so
 by setting up the Hamiltonian, numerically, and diagonalizing it.

 In addition to determining the bound energies, the script also plot
 the corresponding eigen states.

 Numerical inputs:
   L       - The extension of the spatial grid 
   N       - The number of grid points
 
 Input for the confining potential:
   V0      - The "height", the the barrier (must be negative for a well)
   Sigma   - The width of the barrier

 All inputs are hard coded initially.
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt

# Numerical grid parameters
L = 20
N = 512                # Should be 2^k, k integer, for FFT's sake


# Inputs for the smoothly rectangular potential
V0 = -1            # Height, i.e. negative debth
Sigma = np.sqrt(2)          # Width

# Set up grid
x = np.linspace(-L/2, L/2, N)
h = L/(N-1)

# Determine double derivative by means of the fast Fourier transform.
# Set up vector of k-values
k_max = np.pi/h
dk = 2*k_max/N
k = np.append(np.linspace(0, k_max-dk, int(N/2)), 
              np.linspace(-k_max, -dk, int(N/2)))
# Allocate and declare
Tmat = np.zeros((N, N), dtype=complex)
# Transform identity matrix
Tmat = np.fft.fft(np.identity(N, dtype=complex), axis = 0)
# Multiply by (ik)^2
Tmat = np.matmul(np.diag(-k**2), Tmat)
# Transform back to x-representation. 
Tmat = np.fft.ifft(Tmat, axis = 0)
# Correct pre-factor
Tmat = -1/2*Tmat    

# Potential (as function)
def Vpot(x):
    return V0*np.exp(-x**2/(2*Sigma**2))

# Full Hamiltonian
Ham = Tmat + np.diag(Vpot(x))


# Diagaonalize Hamiltonian (Hermitian matrix)
Evector, PsiMat = np.linalg.eigh(Ham)
# Normalize eigenstates
PsiMat = PsiMat/np.sqrt(h)

File: MatlabCode/Chapter8_NonHermitian/Exercise_8_2_4.m
% This script determines the admissible energies for a bound quantum
% particle confined in a rectuangular-like confining potential. It does so
% by setting up the Hamiltonian, numerically, and diagonalizing it.
%
% In addition to determining the bound energies, the script also plot
% the corresponding eigen states.
%
% Numerical inputs:
%   L       - The extension of the spatial grid 
%   N       - The number of grid points
% 
% Input for the confining potential:
%   V0      - The "height" of the potential (must be negative for a well)
%   Sigma   - The width of the potetial
%
% All inputs are hard coded initially.

% Numerical grid parameters
L = 20;
N = 512;                % Should be 2^k, k integer, for FFT's sake

% Input parameters for the potential
V0 = -1;
Sigma = sqrt(2);

% Set up grid
h = L/(N-1);
x = transpose(linspace(-L/2, L/2, N));      % Column vector

% Set up potential
Vpot = @(x) V0*exp(-x.^2/(2*Sigma^2));

% Set up kinetic energy matrix by means of the fast Fourier transform
k=2*pi/(N*h)*[0:(N/2-1), (-N/2:-1)];        % Vector with k-values
% Fourier transform the identity matrix:
Tmat = fft(eye(N));
% Multiply by (ik)^2
Tmat = diag(-k.^2)*Tmat;
% Transform back to x-representation
Tmat = ifft(Tmat);
Tmat = -1/2*Tmat;                   % Correct prefactor

% Full Hamiltonian
Ham = Tmat + diag(Vpot(x));             % Hamiltonian

% Diagonalization
[U E] = eig(Ham);
% Extract eigen energies, the diagonal of E
E = real(diag(E));
% Sort eigenvalues - and rearrange eigenvectors accordingly
[E Ind] = sort(E);
U = U(:, Ind);
% Ensure proper normalization
U = U/sqrt(h);

% Write bound energies to screen and plot wave functions
n = 1;
while E(n) < 0
  % Write energy to screen
  disp(['Energy nr. ', num2str(n),': ', num2str(E(n))])
  % Plot wave function
  plot(x, U(:,n))
  title(['Energy index: ', num2str(n), ', energy: ', num2str(E(n))])
  % Update n
  n = n+1;
end